diagnose_ceph/troubleshoot_ceph.py

```diff
@@ -0,0 +1,285 @@
+        elif (not (cls.options.host and cls.options.user) and
+              not (cls.options.host and cls.options.ssh_key and
+                   cls.options.user)):
+            msg = 'Credentials insufficient, see help'
+            raise SSHCredsNotFoundError(msg)
+
+        if cls.juju_version is 'not_supported':
+            raise JujuInstallationNotFoundError('juju not found/supported')
+        elif cls.juju_version == 'juju1' or cls.juju_version == 'juju2':
+            home = os.path.expanduser('~')
+            if cls.juju_version == 'juju1':
+                cls.pem_location = os.path.join(home, '.juju/ssh/juju_id_rsa')
+            else:
+                loc = '.local/share/juju/ssh/juju_id_rsa'
+                cls.pem_location = os.path.join(home, loc)
+            if not hasattr(cls, 'juju_ceph_machines'):
+                cls.juju_ceph_machines = self._get_all_juju_ceph_machines()
+
+        if cls.options.provider == 'ssh':
+            if not hasattr(cls, 'is_juju'):
+                cls.is_juju = False
+            try:
+                cls.connection = cls._get_connection(cls.options.host)
+            except Exception as err:
+                print err
+                raise ConnectionFailedError('Couldnot connect to host')
+        else:
+            if not hasattr(cls, 'is_juju'):
+                cls.is_juju = True
+        if not hasattr(cls, 'init_type'):
+            cls.init_type = self._get_init_type(cls.connection,
+                                                cls.is_juju).strip()
+        if cls.init_type == 'none':
+            raise InitSystemNotSupportedError()
+        if not hasattr(cls, 'arch_type'):
+            cls.arch_type = self._get_arch_type(cls.connection,
+                                                cls.is_juju).strip()
+
+    def _get_all_machine_param(self, machine):
+        id = machine['machine']
+        public_addr = machine['public-address']
+        cmd = 'juju run --machine ' + str(id) + ' "cat /etc/hostname"'
+        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE)
+
+        hostname = proc.communicate()[0].strip('\n')
+        cmd = 'juju run --machine ' + str(id) + ' "host ' + hostname + ' "'
+        proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE)
+        i_ip = proc.communicate()[0].strip('\n').split(' ')[-1]
+
+        return id, public_addr, hostname, i_ip
+
+    def _get_all_juju_ceph_machines(self):
+        cls = TroubleshootCeph
+        leader_id = sys.maxint
+        cls.connection = None
+
+        proc = subprocess.Popen('juju status --format json', shell=True,
+                                stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        machine_list = json.loads(proc.communicate()[0])
+        juju_machines = []
+        for name, val in machine_list['services']['ceph']['units'].iteritems():
```

Comment:
machine_list should have a check here to make sure ['services']['ceph']['units'] exists or this will blow up